faf-commons-data/src/main/java/com/faforever/commons/replay/ReplaySemantics.java

```diff
@@ -0,0 +1,152 @@
+package com.faforever.commons.replay;
+
+import com.faforever.commons.replay.body.Event;
+import com.faforever.commons.replay.header.Source;
+import com.faforever.commons.replay.shared.LuaData;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ReplaySemantics {
+
+  public static Duration tickToDuration(int tick) {
+    return Duration.ofSeconds(tick / 10);
+  }
+
+  /**
+   * Registers the events by attaching a tick and a source to them.
+   * @param sources
+   * @param events
+   * @return
+   */
+  public static List<RegisteredEvent> registerEvents(List<Source> sources, List<Event> events) {
+    final AtomicInteger tick = new AtomicInteger(0);
+    final AtomicInteger clientId = new AtomicInteger(-1);
+
+    return events.stream().map((event) -> switch (event) {
+      case Event.Advance e -> {
+        tick.addAndGet(e.ticksToAdvance());
+        yield null;
+      }
+
+      case Event.SetCommandSource e -> {
+        clientId.set(e.playerIndex());
+        yield null;
+      }
+
+      default -> new RegisteredEvent(tick.intValue(), sources.get(clientId.intValue()), event);
+    }).filter(Objects::nonNull).toList();
```

Comment:
Looking at this now I might just change this to a normal enhanced for loop that modifies a list since the lambda body modifies the external state so much. Then just a normal int can be used. But I don't know if it is totally required.

Reply:
I think it is nice to be consistent with other solutions that we use to map over the events.